### PR TITLE
Properly separate package entries for APT

### PR DESCRIPTION
**Pull request by marckhouzam**
### What this PR does / why we need it

After publishing the apt package for `v1.0.0-rc.1` we noticed that `v1.0.0-rc.0` was no longer accessible by apt.

After looking at the generated `Packages` files, we realized that there was an empty line missing between the `rc.0` package section and the `rc.1` package section, so the `rc.1` package was overwriting the `rc.0` section.

This PR has the APT automated build insert an empty line between such sections.

However, even if we fix the package build automation script, how to we recover from the already missing two lines (one after `rc.0` and one after `rc.1`) that are already published in our GCP bucket?  We cannot modify them by hand because they are signed.

We've decided that when the script downloads the `Packages` files, it will make sure that there is an empty line between each section.  If there is already an empty line, the script won't need to add one, but if it is missing, it will add the missing empty line.  This is the approach of this PR.

### Which issue(s) this PR fixes

Fixes # N/A

### Describe testing done for PR

Test adding an tanzu-cli-unstable package
```
$ make build-cli-linux-amd64
build linux-amd64 CLI with version: v1.0.0-dev
$ make apt-package
docker run --rm -e VERSION=v1.0.0-dev -e DEB_SIGNER= -v /Users/kmarc/git/tanzu-cli:/Users/kmarc/git/tanzu-cli ubuntu /Users/kmarc/git/tanzu-cli/hack/apt/build_package.sh
++ uname
+ '[' Linux '!=' Linux ']'
++ command -v apt-get
+ PKG_MGR=/usr/bin/apt-get
+ '[' -z /usr/bin/apt-get ']'
+ '[' -z v1.0.0-dev ']'
+ VERSION=1.0.0-dev
+ UNSTABLE=
+ [[ 1.0.0-dev == *-* ]]
+ UNSTABLE=-unstable
[...]
+ curl -sLo /Users/kmarc/git/tanzu-cli/hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages https://storage.googleapis.com/tanzu-cli-os-packages/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
+ cleanPackageFile /Users/kmarc/git/tanzu-cli/hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
+ file=/Users/kmarc/git/tanzu-cli/hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
+ sed '/Description: / a\\' /Users/kmarc/git/tanzu-cli/hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
+ uniq
+ mv /Users/kmarc/git/tanzu-cli/hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages.tmp /Users/kmarc/git/tanzu-cli/hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
[...]
+ echo skip debsigning
skip debsigning

# Verify that the Packages file is now correct
$ cat hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-arm64/Packages
Package: tanzu-cli
Version: 0.90.1
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli/tanzu-cli_0.90.1_arm64.deb
Size: 39490748
SHA256: fe74e6c175e0ca2252cd299e05cf7be1f2027bea96a1923e78ebf447a7f7dfe7
SHA1: 26de20863262c8a1f5adb13cca0057e75225221b
MD5sum: e9b863a8d5f49fd0ed818358aa5c0e75
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.0
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.0_linux_arm64.deb
Size: 39575498
SHA256: 950c25699169d10ac6aab00edda8b8793c4644edc6ef07e201376559d4f9a3ce
SHA1: 7d857e9eeabf31c7b3801e2e96d17e1a0eee3a83
MD5sum: cd30d8c359d60a609a4e8e74062d7247
SHA512: 1b717b6306df1a2d85a0f8b0ddb8409aafb71cea8904889432b5520ae07ac8ae92dc5e41a9faf959a7ec04cdfad0b8cdf422941cd03ed0ee8372284a1c843f8d
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.1
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.1_linux_arm64.deb
Size: 39574894
SHA256: 38fbde6270ebfa7e40e9f6b4f16d4a64c0c61c374a59fa634956a4a519e026f3
SHA1: 1bd1d61f5c42543eb459990ccdd199b20b1c99fc
MD5sum: e7cc310c9b7397ca823d9860310226a7
SHA512: d69eb9d1cb9cdc70d53094556fa1c24bccdcfd3fa397c7f2f851b0806593ca1da807882ea3bbd458a22994c976fa5ddcecc091b867540c2600e12ca822873da8
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-dev
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-dev_linux_arm64.deb
Size: 39578132
SHA256: e85430fa0f7f8a3e63a2dfe2d270ae30236cd80349ab26226505300e7906f64a
SHA1: ad5117feaf2392b577cb907063723d59306cd7aa
MD5sum: b5aa41d3fb7cf2712a84d8d196c43cec
SHA512: d95056cd8d7232b26f24d4fbbb2f7c756eba5e261fab7f1df423e534eca48fb740f85d3906ba287b8fd776ce73b145ae7616ae1dde04ea7d167c31216b1b6e72
Description: The core Tanzu CLI

# Verify the other packages file
$ cat hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
Package: tanzu-cli
Version: 0.90.1
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli/tanzu-cli_0.90.1_amd64.deb
Size: 39490720
SHA256: 7b355acd2fe4ab24eb25ba6b45f6935f12cd7bf187faa4d06e3e57d2f9082dc5
SHA1: 157404cf0e5e16cf8a46762297431ef16d4d6ed1
MD5sum: 9cb9a0a6a9a64dcac330d5e33d26a275
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.0
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.0_linux_amd64.deb
Size: 39575482
SHA256: 8e951a9b29c0f233dc4eb3e798fea933e151ed2d8ce1519502725c719615054a
SHA1: e7b504b848c9bf025caf138c80738c25d18909b6
MD5sum: 4d88e95833c11222acc9bcb6bf1b3740
SHA512: e0cc27ef1a66818af1df4c03c6354f545d67951ba343db34662ae6aec82bdfae3c207b80263fbc82e151a66ae90893ab0fa69e9358bc6f83ff23b48b66511b9e
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.1
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.1_linux_amd64.deb
Size: 39574888
SHA256: 4ef49a2fa10a19087166d33f390ffc5d77d6a2be547d606b44472692325ba49d
SHA1: 6d99a4e869a56004808de971f150738e75f593c2
MD5sum: 90059e16a7dce3c3ec2712f43c687fe8
SHA512: f00817ae48302e9c450cb15866ea58b74fb80e4ff98eb46cabd23c681c90aafe4b9466916b72b70131f5586849694b49d695e719803c19c552dabd7d8520bd20
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-dev
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-dev_linux_amd64.deb
Size: 39578132
SHA256: 3b5f6524d75c2f7bbd9534de8d309334942d4f712f4ab6ba05dad1e8644417cc
SHA1: 48063ecc23d82a865f86d1406138707a2beac9e5
MD5sum: 890351264136387b6948ccef082b6327
SHA512: 8eb7d8fd93838f52ee6dee99d9153710ccf45ce02f5a2749de1d8acf1f8351b27a560c6a5cb07c757c143188950caa6c7124f145071d36ee1446da1a62f650ed
Description: The core Tanzu CLI

# Try to use the repo
$ docker run --rm -it -v $(pwd)/hack/apt/_output/apt:/tmp/apt ubuntu
root@0d68ca772776:/# echo "deb file:///tmp/apt tanzu-cli-jessie main" | tee /etc/apt/sources.list.d/tanzu.list
root@0d68ca772776:/# apt-get update --allow-insecure-repositories
deb file:///tmp/apt tanzu-cli-jessie main
[...]
Get:22 http://ports.ubuntu.com/ubuntu-ports jammy-security/restricted arm64 Packages [517 kB]
Fetched 24.7 MB in 5s (5430 kB/s)
Reading package lists... Done
W: The repository 'file:/tmp/apt tanzu-cli-jessie Release' is not signed.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.
root@0d68ca772776:/# apt list -a tanzu-cli
Listing... Done
tanzu-cli/unknown 0.90.1 arm64

# Notice we now see all tanzu-cli-unstable versions
root@0d68ca772776:/# apt list -a tanzu-cli-unstable
Listing... Done
tanzu-cli-unstable/unknown 1.0.0-rc.1 arm64
tanzu-cli-unstable/unknown 1.0.0-rc.0 arm64
tanzu-cli-unstable/unknown 1.0.0-dev arm64
```

Now test as if we were doing the v1.0.0 release
```
$ git tag v1.0.0
$ make build-cli-linux-amd64
build linux-amd64 CLI with version: v1.0.0
$ make apt-package
docker run --rm -e VERSION=v1.0.0 -e DEB_SIGNER= -v /Users/kmarc/git/tanzu-cli:/Users/kmarc/git/tanzu-cli ubuntu /Users/kmarc/git/tanzu-cli/hack/apt/build_package.sh
++ uname
+ '[' Linux '!=' Linux ']'
++ command -v apt-get
+ PKG_MGR=/usr/bin/apt-get
+ '[' -z /usr/bin/apt-get ']'
+ '[' -z v1.0.0 ']'
+ VERSION=1.0.0
+ UNSTABLE=
+ [[ 1.0.0 == *-* ]]
[...]
+ echo skip debsigning
skip debsigning

$ cat hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-amd64/Packages
Package: tanzu-cli
Version: 0.90.1
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli/tanzu-cli_0.90.1_amd64.deb
Size: 39490720
SHA256: 7b355acd2fe4ab24eb25ba6b45f6935f12cd7bf187faa4d06e3e57d2f9082dc5
SHA1: 157404cf0e5e16cf8a46762297431ef16d4d6ed1
MD5sum: 9cb9a0a6a9a64dcac330d5e33d26a275
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.0
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.0_linux_amd64.deb
Size: 39575482
SHA256: 8e951a9b29c0f233dc4eb3e798fea933e151ed2d8ce1519502725c719615054a
SHA1: e7b504b848c9bf025caf138c80738c25d18909b6
MD5sum: 4d88e95833c11222acc9bcb6bf1b3740
SHA512: e0cc27ef1a66818af1df4c03c6354f545d67951ba343db34662ae6aec82bdfae3c207b80263fbc82e151a66ae90893ab0fa69e9358bc6f83ff23b48b66511b9e
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.1
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.1_linux_amd64.deb
Size: 39574888
SHA256: 4ef49a2fa10a19087166d33f390ffc5d77d6a2be547d606b44472692325ba49d
SHA1: 6d99a4e869a56004808de971f150738e75f593c2
MD5sum: 90059e16a7dce3c3ec2712f43c687fe8
SHA512: f00817ae48302e9c450cb15866ea58b74fb80e4ff98eb46cabd23c681c90aafe4b9466916b72b70131f5586849694b49d695e719803c19c552dabd7d8520bd20
Description: The core Tanzu CLI

Package: tanzu-cli
Version: 1.0.0
Maintainer: Tanzu CLI project team
Architecture: amd64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli/tanzu-cli_1.0.0_linux_amd64.deb
Size: 39578054
SHA256: de5bfc3b09723c78d2a459cb090235a3de47ec6fdbf46ff75b5384aa3e295d13
SHA1: a51b16785cdea2db144bab89233ff45a74e4885c
MD5sum: e5b6a2f8a221b608f1d439fb49af0b31
SHA512: 305bc6c58b4ba425967b29b2c08ef32836a0be2fe0d1fefed496c1491a728554b12063c14ea0dcd32a15fd05cfc934d598b5b88ed7716e06b6541e7556a61f4d
Description: The core Tanzu CLI

$ cat hack/apt/_output/apt/dists/tanzu-cli-jessie/main/binary-arm64/Packages
Package: tanzu-cli
Version: 0.90.1
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli/tanzu-cli_0.90.1_arm64.deb
Size: 39490748
SHA256: fe74e6c175e0ca2252cd299e05cf7be1f2027bea96a1923e78ebf447a7f7dfe7
SHA1: 26de20863262c8a1f5adb13cca0057e75225221b
MD5sum: e9b863a8d5f49fd0ed818358aa5c0e75
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.0
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.0_linux_arm64.deb
Size: 39575498
SHA256: 950c25699169d10ac6aab00edda8b8793c4644edc6ef07e201376559d4f9a3ce
SHA1: 7d857e9eeabf31c7b3801e2e96d17e1a0eee3a83
MD5sum: cd30d8c359d60a609a4e8e74062d7247
SHA512: 1b717b6306df1a2d85a0f8b0ddb8409aafb71cea8904889432b5520ae07ac8ae92dc5e41a9faf959a7ec04cdfad0b8cdf422941cd03ed0ee8372284a1c843f8d
Description: The core Tanzu CLI

Package: tanzu-cli-unstable
Version: 1.0.0-rc.1
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli-unstable/tanzu-cli-unstable_1.0.0-rc.1_linux_arm64.deb
Size: 39574894
SHA256: 38fbde6270ebfa7e40e9f6b4f16d4a64c0c61c374a59fa634956a4a519e026f3
SHA1: 1bd1d61f5c42543eb459990ccdd199b20b1c99fc
MD5sum: e7cc310c9b7397ca823d9860310226a7
SHA512: d69eb9d1cb9cdc70d53094556fa1c24bccdcfd3fa397c7f2f851b0806593ca1da807882ea3bbd458a22994c976fa5ddcecc091b867540c2600e12ca822873da8
Description: The core Tanzu CLI

Package: tanzu-cli
Version: 1.0.0
Maintainer: Tanzu CLI project team
Architecture: arm64
Homepage: https://github.com/vmware-tanzu/tanzu-cli
Priority: optional
Section: main
Filename: pool/main/t/tanzu-cli/tanzu-cli_1.0.0_linux_arm64.deb
Size: 39578054
SHA256: 551930632b9aa557ee18df8ec7e689528c3833f8504e0d61ec86bb7835010b23
SHA1: 25159ea123420c1afc30b7f8c3e2e168744c5d42
MD5sum: 083c0390a03e16093a1f3892dc57fe33
SHA512: c0097422d14d78e426a01d32489e920a56d314fcdf8444815c6055e494cfbedf5bc62900aa9b138a13ddac3a7dd3a62c1f1e36e2441125f248613228fa4cc3d9
Description: The core Tanzu CLI

$ docker run --rm -it -v $(pwd)/hack/apt/_output/apt:/tmp/apt ubuntu
root@a367b27cf47d:/# echo "deb file:///tmp/apt tanzu-cli-jessie main" | tee /etc/apt/sources.list.d/tanzu.list
root@a367b27cf47d:/# apt-get update --allow-insecure-repositories
deb file:///tmp/apt tanzu-cli-jessie main
[...]
Get:22 http://ports.ubuntu.com/ubuntu-ports jammy-security/restricted arm64 Packages [517 kB]
Fetched 24.7 MB in 4s (5722 kB/s)
Reading package lists... Done
W: The repository 'file:/tmp/apt tanzu-cli-jessie Release' is not signed.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.

# We see multiple tanzu-cli packages
root@a367b27cf47d:/# apt list -a tanzu-cli
Listing... Done
tanzu-cli/unknown 1.0.0 arm64
tanzu-cli/unknown 0.90.1 arm64

# We see multiple tanzu-cli-unstable packages
root@a367b27cf47d:/# apt list -a tanzu-cli-unstable
Listing... Done
tanzu-cli-unstable/unknown 1.0.0-rc.1 arm64
tanzu-cli-unstable/unknown 1.0.0-rc.0 arm64

# Test that we can install the v1.0.0 version
root@a367b27cf47d:/# apt install tanzu-cli
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  tanzu-cli
0 upgraded, 1 newly installed, 0 to remove and 4 not upgraded.
Need to get 0 B/39.6 MB of archives.
After this operation, 0 B of additional disk space will be used.
WARNING: The following packages cannot be authenticated!
  tanzu-cli
Install these packages without verification? [y/N] y
Get:1 file:/tmp/apt tanzu-cli-jessie/main arm64 tanzu-cli arm64 1.0.0 [39.6 MB]
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package tanzu-cli.
(Reading database ... 4389 files and directories currently installed.)
Preparing to unpack .../tanzu-cli_1.0.0_linux_arm64.deb ...
Unpacking tanzu-cli (1.0.0) ...
Setting up tanzu-cli (1.0.0) ...

root@a367b27cf47d:/# tanzu version
version: v1.0.0
buildDate: 2023-08-03
sha: ca7ba9c8a
```

### Release note

```release-note
Properly separate the metadata section of different apt package version.
```

### CODE REVIEW

**Review by anujc25 — APPROVED** [review]
LGTM

File: hack/apt/build_package_repo.sh [hunk 1]
```diff
@@ -42,6 +42,21 @@ if ! command -v bzip2 &> /dev/null || ! command -v curl &> /dev/null; then
    ${PKG_MGR} install -y bzip2 curl
 fi
 
+function cleanPackageFile {
+   file=$1
+
+   # There must be an empty line between each package section.
+   # Our package sections alwaya end with the "Description: " line,
```
Comment (vuil) [inline hack/apt/build_package_repo.sh:49]: alwaya->always

File: hack/apt/build_package_repo.sh [hunk 2]
```diff
@@ -83,6 +100,7 @@ SHA1: $(sha1sum ${PKG} | cut -f1 -d' ')
 MD5sum: $(md5sum ${PKG} | cut -f1 -d' ')
 SHA512: $(sha512sum ${PKG} | cut -f1 -d' ')
 Description: The core Tanzu CLI
+
```
Comment (vuil) [inline hack/apt/build_package_repo.sh:103]: nit: don't think it matters anymore now that we have the cleanser, but would the right thing be to add a blank line here?

**Review by vuil — COMMENTED** [review]
lgtm, just a few small things

File: hack/apt/build_package_repo.sh [hunk 3]
```diff
@@ -42,6 +42,21 @@ if ! command -v bzip2 &> /dev/null || ! command -v curl &> /dev/null; then
    ${PKG_MGR} install -y bzip2 curl
 fi
 
+function cleanPackageFile {
+   file=$1
+
+   # There must be an empty line between each package section.
+   # Our package sections alwaya end with the "Description: " line,
+   # so let's append an empty line after that Description line.
```
Comment (vuil) [inline hack/apt/build_package_repo.sh:50]: > let's append an empty line after that Description line.

 strictly this is more about appending an empty line after EVERY "Description: " line.

File: hack/apt/build_package_repo.sh (hunk 2 above)
Comment (marckhouzam) [reply hack/apt/build_package_repo.sh:103]: Yes.  But it is added, right?  Empty "+" line.

Comment (vuil) [reply hack/apt/build_package_repo.sh:103]: duh, your are right.

**Pull request merged**